### ISSUE TITLE
distributed training is quite slower than standalone training

### ISSUE THREAD

**Issue by tjuzhao111** [dmlc/mxnet#2030]
I have modified the original mnist example for using my own numpy data. The code run on 3 machines successfully. Each machine read one part of the whole training data. Every machine have 2 cpus which were all used.

When I run the code standalone in one machine without kv-store, it's process speed is:
`2016-05-04 20:39:30,225 Node[0] Epoch[0] Batch [500]	Speed: 106774.23 samples/sec	Train-accuracy=0.857792`
`2016-05-04 20:39:43,554 Node[0] Epoch[0] Batch [1000]	Speed: 112538.91 samples/sec	Train-accuracy=0.852428`

But when I run it on 3 machines simultaneously with mpi, the processing speed became 10 times slower than standalone running.
`2016-05-04 21:04:18,278 Node[2] Epoch[0] Batch [500]	Speed: 12586.55 samples/sec	Train-accuracy=0.837458`
`2016-05-04 21:04:18,367 Node[0] Epoch[0] Batch [500]	Speed: 12577.49 samples/sec	Train-accuracy=0.857792`
`2016-05-04 21:04:18,409 Node[1] Epoch[0] Batch [500]	Speed: 12580.81 samples/sec	Train-accuracy=0.847393`

I am really confused now. Can you help me?

**Comment by piiswrong** [comment]
MNIST is too small. Communication cost dominates computation. 
Try inception

**Comment by tjuzhao111** [comment]
@piiswrong Thank you for your reply. I use my own data which has 136359166 samples with 24 dimensions. The batchsize is 3000.  

I have observed more experiments results. When I run the MNIST example, I fonud that the running speed doesn't decrease.
`local`
`2016-05-04 20:02:28,642 Node[0] Epoch[0] Batch [50]	Speed: 501.89 samples/sec	Train-accuracy=0.686719`
`2016-05-04 20:02:41,445 Node[0] Epoch[0] Batch [100]	Speed: 499.90 samples/sec	Train-accuracy=0.900937`

`dist_sync 3 machines`
`2016-05-04 20:05:54,708 Node[0] Epoch[0] Batch [50]	Speed: 480.65 samples/sec	Train-accuracy=0.716562`
`2016-05-04 20:05:54,710 Node[1] Epoch[0] Batch [50]	Speed: 479.89 samples/sec	Train-accuracy=0.712500`
`2016-05-04 20:05:54,711 Node[2] Epoch[0] Batch [50]	Speed: 480.72 samples/sec	Train-accuracy=0.718437`

`dist_async 3 machines`
`2016-05-04 20:08:25,341 Node[0] Epoch[0] Batch [50]	Speed: 403.37 samples/sec	Train-accuracy=0.188594`
`2016-05-04 20:08:25,965 Node[1] Epoch[0] Batch [50]	Speed: 388.19 samples/sec	Train-accuracy=0.180469`
`2016-05-04 20:08:25,967 Node[2] Epoch[0] Batch [50]	Speed: 388.25 samples/sec	Train-accuracy=0.189688`

It looks like normal. But when I use NDArrayIter, the running speed decreased heavily. Then I changed the batchsize from 3000 to 10000. The speed recovered a little.
`with batch size 10000`
`2016-05-05 10:32:59,625 Node[2] Epoch[0] Batch [500]	Speed: 28058.05 samples/sec	Train-accuracy=0.837400`
`2016-05-05 10:32:59,797 Node[0] Epoch[0] Batch [500]	Speed: 28058.64 samples/sec	Train-accuracy=0.854208`
`2016-05-05 10:32:59,876 Node[1] Epoch[0] Batch [500]	Speed: 28028.72 samples/sec	Train-accuracy=0.852116`

`batch size 20000`
`2016-05-05 10:57:26,121 Node[2] Epoch[0] Batch [500]	Speed: 45233.51 samples/sec	Train-accuracy=0.834638`
`2016-05-05 10:57:26,288 Node[1] Epoch[0] Batch [500]	Speed: 45235.94 samples/sec	Train-accuracy=0.854388`
`2016-05-05 10:57:26,338 Node[0] Epoch[0] Batch [500]	Speed: 45213.37 samples/sec	Train-accuracy=0.854408`

I have a doubt that the bottleneck is the NDArrayIter. Maybe I need to modified the c++ source code to improve this situation?

Or do you have any support for numerical dense/sparse feature vector training samples?

**Comment by tqchen** [comment]
You can simply try to run a dummy iterator that always output the same batch. to see if the data loading is the problem.
